---
int32[] points_array
int32[] shape